enterprise/frontend/src/embedding-sdk/hooks/private/use-init-data.ts

```diff
@@ -36,65 +37,76 @@ const getErrorMessage = (authType: SDKConfigType["authType"]) => {
   const dispatch = useSdkDispatch();
 
   const loginStatus = useSdkSelector(getLoginStatus);
 
-  const [sessionTokenState, setSessionTokenState] =
-    useState<EmbeddingSessionTokenState | null>(null);
-
   useEffect(() => {
     registerVisualizationsOnce();
     dispatch(setLoginStatus({ status: "uninitialized" }));
   }, [dispatch]);
 
-  const jwtProviderUri =
-    config.authType === "jwt" ? config.jwtProviderUri : null;
+  const setupJwtAuth = useCallback(
+    (config: SdkConfigWithJWT) => {
+      api.onBeforeRequest = async () => {
+        const tokenState = await dispatch(
+          getOrRefreshSession(config.jwtProviderUri),
+        );
 
-  useEffect(() => {
-    if (config.authType === "jwt") {
-      const updateToken = () => {
-        const currentState = store.getState();
-        setSessionTokenState(getSessionTokenState(currentState));
+        api.sessionToken = (
+          tokenState.payload as EmbeddingSessionTokenState["token"]
+        )?.id;
       };
-
-      const unsubscribe = store.subscribe(updateToken);
-
-      if (jwtProviderUri) {
-        dispatch(getOrRefreshSession(jwtProviderUri));
-      }
-
-      updateToken();
-
-      return () => unsubscribe();
-    }
-  }, [config.authType, dispatch, jwtProviderUri]);
-
-  useEffect(() => {
-    api.basename = config.metabaseInstanceUrl;
-
-    if (config.authType === "jwt" && config.jwtProviderUri) {
-      api.onBeforeRequest = () =>
-        dispatch(getOrRefreshSession(config.jwtProviderUri));
-      api.sessionToken = sessionTokenState?.token?.id;
       dispatch(setLoginStatus({ status: "initialized" }));
-    } else if (config.authType === "apiKey" && config.apiKey) {
+    },
+    [dispatch],
+  );
+
+  const setupApiKeyAuth = useCallback(
+    (config: SdkConfigWithApiKey) => {
       api.apiKey = config.apiKey;
       dispatch(setLoginStatus({ status: "initialized" }));
-    } else {
-      dispatch(
-        setLoginStatus({
-          status: "error",
-          error: new Error(getErrorMessage(config.authType)),
-        }),
-      );
+    },
+    [dispatch],
+  );
+
+  useEffect(() => {
+    if (loginStatus.status === "uninitialized") {
+      api.basename = config.metabaseInstanceUrl;
+
+      if (isJwtAuth(config)) {
+        setupJwtAuth(config);
+      } else if (isApiKeyAuth(config)) {
+        setupApiKeyAuth(config);
+      } else {
+        dispatch(
+          setLoginStatus({
+            status: "error",
+            error: new Error(getErrorMessage(config.authType)),
+          }),
+        );
+      }
     }
-  }, [config, dispatch, sessionTokenState]);
+  }, [
+    config,
+    config.authType,
+    config.metabaseInstanceUrl,
+    dispatch,
+    loginStatus.status,
+    setupApiKeyAuth,
+    setupJwtAuth,
+  ]);
 
   useEffect(() => {
-    const fetchData = async () => {
-      if (loginStatus.status === "initialized") {
+    if (loginStatus.status === "initialized") {
```

Comment:
I want to ensure that this only runs once data is validated and loaded. (On another note, maybe the `initialized` status could be named to `validated`)

Reply:
I actually like `validated` more than `initialized` I think in some systems, we might even consider `initialized` as a ready state for the application, but it's not the case for us.

Reply:
Nice. Renamed the state, it feels right